designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
+Success criteria for the change
+-------------------------------
+#. Support all programming languages supported by AWS Lambda
+
+   * Nodejs with NPM
+   * Java with Maven
+   * Python with PIP
+   * Golang with Go CLI
+   * Dotnetcore with DotNet CLI
+
+
+#. Each Lambda function in SAM template gets built
+
+#. Produce stable builds (best effort): If the source files did not change, built artifacts should not change.
+
+#. Built artifacts should "just work" with ``sam local`` and ``sam package`` suite of commands
+
+#. Ability to provide arbitrary build commands for each Lambda Function runtime. This will either override the built-in
+   default or add build support for languages/tools that SAM CLI does not support (ex: java+gradle).
+
+#. Opt-in to building native dependencies that can run on AWS Lambda using Docker.
+
+#. Support one dependency manifest (ex: package.json) entire app or one per each Lambda function.
+
+#. Support out-of-source builds: ie. source code of Lambda function is outside the directory containing SAM template.
+
+#. Integrate build action with ``sam local/package/deploy`` commands so the Lambda functions will be automatically
+   built as part of the command without explicitly running the build command.
+
+
+Out-of-Scope
+------------
+#. Supports adding data files ie. files that are not referenced by the package manager (ex: images, css etc)
+#. Support to exclude certain files from the built artifact (ex: using .gitignore or using regex)
+#. If the app contains a ``buildspec.yaml``, automatically run it using CodeBuild Local.
+#. Watch for file changes and build automatically (ex: ``sam build --watch``)
+#. Support other build systems by default Webpack, Yarn or Gradle.
+#. Support in AWS CLI, ``aws cloudformation``, suite of commands
+#. Support for fine-grained hooks (ex: hooks that run pre-build, post-build, etc)
+
+
+User Experience Walkthrough
+---------------------------
+Let's assume customers has the following SAM template:
+
+    **NOTE**: Currently we advice customers to set *CodeUri* to a folder containing built artifacts that can be readily
+    packaged by ``sam package`` command. But to use with *build* command, customers need to set *CodeUri* to the folder
+    containing source code and not built artifacts.
+
+.. code-block:: yaml
+
+    MyFunction1:
+        Type: AWS::Lambda::Function
+        Properties:
+            ...
+            Code: ./source-code1
+            ...
+
+    MyFunction2:
+        Type: AWS::Serverless::Function
+        Properties:
+            ...
+            Code: ./source-code2
+            ...
+
+
+To build, package and deploy this app, customers would do the following:
+
+**1. Build:** Run the following command to build all functions in the template and output a SAM template that can be run through
+the package command:
+
+.. code-block:: bash
+
+    # Build the code and write artifacts to ./build folder
+    $ sam build -t template.yaml -b ./build -o built-template.yaml
```

Comment:
For a better user experience, I'd like to suggest adding them as default values and have the user to simply run `sam build`. 

Flags would be then used to for custom scenarios of which they can also use `samrc` config file

Reply:
Agreed. I am showing them here for explanation, but all of these will have defaults so you type nothing but `sam build`